### PR TITLE
Occurrence Updates

### PR DESCRIPTION
**Pull request by roleyfoley**
## Description
Adds a couple of changes to the occurrence and its processing 
- Adds an occurrence cache to reduce the repetition of occurrence building during link processing  
- Moves the setting namespace configuration for the occurrence into it's own section of the occurrence. This keeps the settings lookups consistent and also allows them to be visible to users without having to modify the hamlet code itself ( with debug statements ) 

## Motivation and Context
Occurrence caching is for performance reasons 
SettingNamespace is for better user understanding

## How Has This Been Tested?
Tested locally 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions
- [x] None

## Checklist:
- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] None of the above.

### CODE REVIEW

File: engine/occurrence.ftl [hunk 1]
```diff
@@ -461,6 +515,25 @@
     [#return result]
 [/#function]
 
+[#function internalGetOccurrenceSettingNamepsaces occurrence ]
```
Comment (RossMurr4y) [inline engine/occurrence.ftl:518]: typo here in `Namespaces`

File: engine/occurrence.ftl [hunk 2]
```diff
@@ -167,6 +208,19 @@
             }
         ) ]
 
+    [#-- define the setting namespaces for the component --]
+    [#local occurrence =
+        mergeObjects(
+            occurrence,
+            {
+                "Configuration" : {
+                    "SettingNamespaces" : internalGetOccurrenceSettingNamepsaces(occurrence)
```
Comment (RossMurr4y) [inline engine/occurrence.ftl:217]: Type here in `Namespaces`, copied from the function definition I'd say

**Comment by RossMurr4y** [comment]
Looks good, should really speed things up.

Had a couple questions that I couldn't see covered by the code:

1) How does this handle an existing cache but altered solution? If I update my solution I'd expect it not to retrieve the solution from cache, or perhaps the cache to be flushed because I updated the solution.

2) With the cache now in the user home directory, during CI builds, wont the cache always exist past the deployment unit's first run?

File: engine/occurrence.ftl (hunk 1 above)
Comment (roleyfoley) [reply engine/occurrence.ftl]: Fixed

File: engine/occurrence.ftl (hunk 2 above)
Comment (roleyfoley) [reply engine/occurrence.ftl]: Fixed

**Comment by roleyfoley** [comment]
> Looks good, should really speed things up.
> 
> Had a couple questions that I couldn't see covered by the code:
> 
>     1. How does this handle an existing cache but altered solution? If I update my solution I'd expect it not to retrieve the solution from cache, or perhaps the cache to be flushed because I updated the solution.
> 
>     2. With the cache now in the user home directory, during CI builds, wont the cache always exist past the deployment unit's first run?

This doesn't cache outside of a given freemarker pass, it only caches occurrences which are called in the same pass multiple times 

Caching to a file would require us to be able to write multiple files at once. Something we are working on as part of the freemarker wrapper. That would be a significant performance increase over what is there now

**Review by ml019 — APPROVED** [review]
No Issues. Making namespaces part of occurrence is a nice optimisation.

**Pull request merged**